만국 박람회 [Step2] 우롱차, marisol

dnwhd0112 opened the pull request:
@lina0322
안녕하세요 엘림 이번 Pr도 잘 부탁드려요

## 고민한점 & 해결한 방법

### 1. 첫번째 화면에서 두 번째 화면으로 전환할 때도 네비게이션 컨트롤러로 이동해야 할까?

첫번째 화면을 스크롤하면 상단에 네비게이션 바 영역이 없어서 네비게이션 컨트롤러를 두 번째 화면에 붙여야 하나?! 라고 생각했습니다.
하지만 두번째 화면에서 첫번째 화면으로 돌아올때 Navigation Bar Item을 통해서 돌아와야 해서, 네비게이션 컨트롤러를 첫번째 화면에 embed 하고, `viewWillAppear()`에서 `navigationController?.setNavigationBarHidden()`메서드를 이용해 네비게이션 바 영역을 숨겨주었습니다.

처음에는 viewDidLoad에 해당 메서드를 호출했는데, 가장 처음에 한번만 실행되고, 다음 화면에 갔다가 돌아오면 다시 네비게이션 바 영역이 살아나서 viewWillAppear에서 해당 메서드를 호출하도록 했습니다.


### 2. 스토리 보드에서 스크롤 하기 위하여 constraint를 주는게 미숙하여 많이 삽질했습니다. 
-> 해결방법 더 많은 삽질로 해결했습니다.

### 3. 뷰간 전환할때 data를 넘기는 방법에 대하여 고민하였습니다.
 처음에는 `delegate패턴`을 적용할려고 하였으나 구현해보니 적절치 않은 방법이라 생각되어 다른 방법을 고민하였습니다. 그래서 `viewController`에서 변수를 할당하여 뷰간 전환할때 해당 변수에 값을 넣어주는 방법으로 구현하였습니다. 
```swift=
guard let detailViewController = storyboard?.instantiateViewController(withIdentifier: IdentifierName.detailViewController) as? DetailViewController else {
            return
        }
        
        detailViewController.item = items[indexPath.row]
        self.navigationController?.pushViewController(detailViewController, animated: true)
```
하지만 이러한 방법은 view 의 변수가 외부에서 바뀔수 있다는 단점이 있습니다. 때문에 웨더가 다른방법을 알려주었습니다.(이니셜라이져로 값을 할당하는 방식)

```swift=
    let detailViewController = UIStoryboard(name: IdentifierName.detailViewController, bundle: .main).instantiateInitialViewController { coder -> DetailViewController? in
            DetailViewController(coder: coder, item: self.items[indexPath.row])
        }
    
```
하지만 저희가 이 코드를 완벽히 이해하지 못한다고 판단하여 코드에 적용하지는 않았습니다.

### 4. 메인화면에서 한 줄인 문자열을 어떻게 두 줄로 나타낼까?
json 데이터에서 “파리 만국박람회 1900 (L’Exposition de Paris 1900)”으로 한 줄로 되어 있는 데이터를 괄호를 기준으로 두 줄로 표시하는 방법을 고민했습니다. 
고민하다가 `replacingOccurrences(of:, with: )` 메서드를 통해 아래와 같이 괄호”(“를 찾아 줄바꿈 해주는 문자열을 추가한 괄호로 바꿔주었습니다.

`title.replacingOccurrences(of: “(“, with: “\n(“)`

혹시 다른 방법도 있을지 궁금합니다!

### 5. UITableViewCell을 리턴하는 메서드의 옵셔널 바인딩 처리
출품작 리스트를 보여주는 HeritageListViewController에서 datasource 부분에서 cell을 생성하는 메서드가 있습니다.
`func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell`
이 메서드에서 guard let 옵셔널 바인딩을 해줄 때, 반드시 return 값이 있어야 하는데 return nil을 할 수 없어서 
오류 발생시 `return UITableViewCell()`로 빈 cell을 return 하도록 했습니다.😭
옵셔널이 아닌 값을 return 해야 할 때 이런 식으로 빈 cell을 return 해줘도 되는건지, 더 좋은 방법이 있을지 궁금합니다!

dnwhd0112 committed:
feat: 스토리보드 구현

marisol-develop committed:
feat: MainViewController구현

- loadInformation 구현
- setContentsOfScrollView 구현
- editTitle 구현

dnwhd0112 committed:
feat: 두번째 화면구현

marisol-develop committed:
feat: HeritageListViewController 구현

- loadItems 구현
- UITableViewDataSource 구현

dnwhd0112 committed:
refactor: 두번째 화면 수정

dnwhd0112 committed:
feat: 3번째 화면 구현

marisol-develop committed:
feat: DetailViewController 구현

- setContentsOfScrollView 구현

dnwhd0112 committed:
refactor: 코드 가독성 개선 및 함수들 private으로 변경

marisol-develop committed:
refactor: 메서드 네이밍 수정 및 주석 수정

dnwhd0112 committed:
refactor: 캡슐화 및 변경 메소드 위치 변경

lina0322 commented on Expo1900/DetailViewController.swift:
```diff
@@ -0,0 +1,33 @@
+
+// MARK: - logic method
+extension DetailViewController {
+    private func setTextOfDetailView() {
```
이미 디테일뷰안에서 이뤄지는 작업이니까, `setText`도 나쁘지 않겠네요.
근데 이미지도 세팅하는 것 같으네요?!

marisol-develop replied:
텍스트와 이미지를 모두 포함하기 위해서 
`setContentsOfDetailView`로 메서드명 수정했습니다!

lina0322 commented on Expo1900/Expo1900.xcodeproj/project.pbxproj:
```diff
@@ -9,12 +9,16 @@
 		C79FF4B52589F401005FB0FD /* AppDelegate.swift in Sources */ = {isa = PBXBuildFile; fileRef = C79FF4B42589F401005FB0FD /* AppDelegate.swift */; };
 		C79FF4B72589F401005FB0FD /* SceneDelegate.swift in Sources */ = {isa = PBXBuildFile; fileRef = C79FF4B62589F401005FB0FD /* SceneDelegate.swift */; };
-		C79FF4B92589F401005FB0FD /* ViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = C79FF4B82589F401005FB0FD /* ViewController.swift */; };
+		C79FF4B92589F401005FB0FD /* MainViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = C79FF4B82589F401005FB0FD /* MainViewController.swift */; };
```
명확한 이름이네요👍

lina0322 commented on Expo1900/HeritageListCell.swift:
```diff
@@ -0,0 +1,29 @@
+
+    override func setSelected(_ selected: Bool, animated: Bool) {
+        super.setSelected(selected, animated: animated)
+    }
```
override를 하는데 super를 그대로 호출하고 추가된 동작이 아무것도 없네요👀
override는 무엇인가요? super에서 호출하는 것은 어떤 의미를 가지고 있죠?
상속에 대해서 설명해주실 수 있나요🙏

marisol-develop replied:
override는 부모 타입에서 사용된 함수를 갖고와서 수정해서 사용하기 위해 쓰는 키워드입니다.
super에서 호출하면 부모 타입에서 갖고 있던 기능을 그대로 가져와 쓴다는 의미로 알고 있습니다.
상속은 부모 타입의 프로퍼티, 메서드들을 그대로 물려받아 사용할 수 있는 기능입니다.

`UITableViewCell`을 상속하는 `HeritageListCell`를 생성하니 자동으로 `awakeFromNib`과 `setSelected`가 생성되어서, 그대로 PR을 했었습니다.🥲

`awakeFromNib`
- interface builder에서 객체가 초기화된 후 호출됩니다
- 연결된 객체의 변수를 사용하려고 할 때 사용됩니다
- 실험 결과 `viewDidLoad` 이전에 사용됩니다 (`awakeFromNib` -> `viewDidLoad` -> `viewWillAppear` 순)
- 또한 IBoutlet의 인스턴스 생성 및 할당은 viewDidLoad 이전에 이뤄집니다

`setSelected(_ selected: Bool, animated: Bool) `
- cell의 선택된 상태가 바뀔 때마다 호출되며, didSelected나 didDeselected로 조건을 걸어 원하는 시점에 코드를 실행할 수 있습니다
- cell이 load 될때마다 호출됩니다. 이때 selected는 false 값으로 호출됩니다.

공부를 해보니, 저희 프로젝트에서 저 두 메서드에 따로 구현을 해줄 사항이 없다는 생각이 들어서, 두 메서드는 삭제하도록 했습니다.

lina0322 commented on Expo1900/HeritageListCell.swift:
```diff
@@ -0,0 +1,29 @@
+    func setContentOfCell(_ item: Item) {
+        self.heritageImage.image = UIImage(named: item.imageName)
+        self.heritageTitleLabel.text = item.name
+        self.heritageShortDescriptionLabel.text = item.shortDescription
```
어디는 self를 명시하고, 어디는 하지 않고있네요.
어떤 기준으로 작업하는지 혼란스러워요🥲

marisol-develop replied:
self를 정확한 기준 없이 사용하고 있었네요 🥲
인스턴스 내부의 프로퍼티와 변수가 네이밍이 같은 경우, 구분을 위해 프로퍼티 앞에 self를 붙이는 기준을 만들고, 나머지 경우에는 self를 삭제했습니다.

lina0322 commented on Expo1900/HeritageListViewController.swift:
```diff
@@ -0,0 +1,77 @@
+        self.navigationController?.navigationBar.topItem?.title = "메인"
+        self.title = "한국의 출품작"
+        heritageListTableView.dataSource = self
+        heritageListTableView.delegate = self
```
셋팅하는 내용들이네요. viewDidLoad에 이렇게 매번 한줄씩 추가하는게 좋을까요? 함수로 빼는게 좋을까요?😉
cell에서 했던 방식을 생각해봅시답

marisol-develop replied:
초기 세팅을 해주는 `setInitialSetting` 메서드를 따로 만들었습니다!

lina0322 commented on Expo1900/HeritageListViewController.swift:
```diff
@@ -0,0 +1,77 @@
+        return items.count
+    }
+    
+    func numberOfSections(in tableView: UITableView) -> Int {
+        return 1
+    }
```
numberOfSections함수의 default return값이 정해져있을까요 아닐까요?👀

marisol-develop replied:
엘림께서 말씀해주셔서 알게 되었는데, 이 메서드를 실행시키지 않으면 자동으로 table은 섹션을 1로 설정하네요!

lina0322 commented on Expo1900/HeritageListViewController.swift:
```diff
@@ -0,0 +1,77 @@
+    
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        navigationController?.setNavigationBarHidden(false, animated: true)
```
`navigationController?.navigationBar.isHidden = false`랑 어떤 차이가 있을까요??

marisol-develop replied:
`isHidden`은 프로퍼티고, `setNavigationBarHidden(_:animated:)`는 메서드입니다.
`setNavigationBarHidden`은 파라미터로 Bool 타입인 animated을 갖으며, animation 설정이 가능하나, `isHidden`은 Bool값을 통해 숨기거나 나타나게 할 수만 있고, animation 설정은 못해주는 것 같아요!

lina0322 commented on Expo1900/HeritageListViewController.swift:
```diff
@@ -0,0 +1,77 @@
+            return UITableViewCell()
+        }
+        
+        let item = items[indexPath.row]
```
index를 벗어나는 값이 들어가면 터지겠네용ㅎㅎ 
로직상으로 그럴일이 없어보이지만, 노올랍게도 그런일이 발생하기도 한답니다..?
방지해줘도 좋고, 감이 안 온다면 나중에 해봐도 좋아요😀

marisol-develop replied:
Array의 extension으로 subscript를 구현해보았습니다.
index가 반드시 array의 indices 범위 내에 존재해야 self[index]를 리턴하고, 아닐 경우 nil이 리턴되도록 했습니다.

```swift
extension Array {
    public subscript(safe index: Int) -> Element? {
        return indices ~= index ? self[index] : nil
    }
}
```

dnwhd0112 replied:
extension으로 값을 벗어난 걸 집어넣으면 nil을 반환하도록 수정했습니다!

lina0322 commented on Expo1900/HeritageListViewController.swift:
```diff
@@ -0,0 +1,77 @@
+        }
+        
+        let item = items[indexPath.row]
+        cell.setContentOfCell(item)
```
cell이 모든걸 담당하네요~!
GOOD👏

lina0322 commented on Expo1900/HeritageListViewController.swift:
```diff
@@ -0,0 +1,77 @@
+    }
+}
+
+// MARK: - UITableViewDelegate method
```
그냥 별겨아닌데.. method라고 명시해주는게 의미가 있나 했어요 ㅎㅎ...

lina0322 commented on Expo1900/HeritageListViewController.swift:
```diff
@@ -0,0 +1,77 @@
+        }
+        
+        detailViewController.item = items[indexPath.row]
+        self.navigationController?.pushViewController(detailViewController, animated: true)
```
self를 제거하면 어떻게되나요? 왜 그런가요?

dnwhd0112 replied:
self를 제거해도 정상적으로 됩니다. class에서 상속된 프러퍼티라 그런거 같아요!

lina0322 replied:
`class에서 상속된 프로퍼티`라는게 어떤 의미일까용?

dnwhd0112 replied:
HeritageListViewController가 UIViewController를 상속 받는대 UIViewController에서 navigationController이 구현되어 있기때문에 사용해도 문제가 없습니다.
self를 사용할때는 다음과 같을때만 쓰기로 마리솔과 합의했습니다.
self는 모든 인스턴스들이 암시적으로 생성하는 프로퍼티이다. 인스턴스 내부의 프로퍼티와 변수가 이름이 같은 경우, 구분을 위해 프로퍼티 앞에 self를 붙인다

lina0322 commented on Expo1900/HeritageListViewController.swift:
```diff
@@ -0,0 +1,77 @@
+            let heritageInfo = try jsonManager.decodedItems()
+            return heritageInfo
+        } catch {
+            print(error.localizedDescription)
```
print는 평소에 디버깅할때는 좋지만, 빌드하는 곳에는 사용하지 않는게 좋아요.
왜인지, 어떤걸로 대체할 수 있는지 찾아볼까요?🙏
또한 이 경우에는 화면에 아무것도 안 뜨겠네요? 사용자는 그럼 어떻게 생각할까요🤔

dnwhd0112 replied:
프로토콜로 error을 받았을때 Alert을 반환해주는 메소드를 구현하여 처리하였습니다. 그래서 사용자가 오류가 발생하면 알람창을 볼수 있도록 수정했습니다!

lina0322 commented on Expo1900/IdentifierName.swift:
```diff
@@ -0,0 +1,12 @@
+//
+
+enum IdentifierName {
+    static let heritageCell = "HeritageCell"
+    static let heritageListViewController = "HeritageListViewController"
+    static let detailViewController = "DetailViewController"
```
뷰컨 이름을 변경하면 여기와서 바꿔야겠네용ㅎㅎ
저들이 자체적으로 이름을 들고있는 방법은 없을까요?😀

dnwhd0112 replied:
열심히 찾아보았는대 도저히 모르겠습니다... 나중에 더 찾아볼게요! 힌트 하나만 주시면 안될까요?  😉

lina0322 replied:
`String(describing: self)` 를 통해서 자신의 이름을 반환할 수 있어요😉
힌트가 조금 되었나요??

dnwhd0112 replied:
넵 적용해보겠습니다!

marisol-develop replied:
아래와 같이 identifier에 자신의 이름을 넣도록 했습니다!
`let identifier = String(describing: HeritageListCell.self)`

lina0322 commented on Expo1900/JsonManager.swift:
```diff
@@ -41,3 +41,34 @@ struct JsonManager: JsonManagerable {
 }
+
+
+struct JsonManager2: JsonManagerable {
```
2탄이 왜 필요했죠ㅠㅠ.. 네이밍이 너어어무 아쉬운데 이게 무슨일인가요😱

dnwhd0112 replied:
이거는 실수로 넣은 코드입니다 ㅠㅠㅠㅠ
마리솔에게 JsonManagerble을 채택해서 구현했을때 이점을 설명할때 썻던 코드에요.
프로토콜로 채택한 경우 변경점이 생겼을때 기존 코드를 적게 변경하면서 사용할 수 있다고 설명한 예시였습니다.
만일 다른곳에서도 JsonManagerble 프로토콜을 사용하였을때 디코딩 하는 방식이 변경되면 기존 코드를 적게 변경하면서 새로운 타입에서 해당 프로토콜을 채택하는 방식으로 구현하면 적게 변경할 수 있다고 설명하면서 쓴 코드입니다. ㅎㅎㅎ 못본척 해주시면 감사하겠습니다😭

lina0322 commented on Expo1900/JsonManager.swift:
```diff
@@ -41,3 +41,34 @@ struct JsonManager: JsonManagerable {
+        }
+        
+        return expositionInfo
+    }
```
코드가 상당히 유사해보이네요ㅎㅎ
당장 수정하기에는 너무 어려운 수준일 것 같지만, 고민해보는 것도 좋을 것 같아요
해결은 시간이 지나면 자연스럽게 할 수 있게 될것도 같네요

dnwhd0112 replied:
제네릭으로 합칠려고 했으나 수정해야 할 곳도 많고 잘 되지가 않더라구요. 일단 메서드 오버로딩으로 변경해 놓았습니다. ㅠㅠ 이부분은 조금 더 해볼게요!

lina0322 commented on Expo1900/MainViewController.swift:
```diff
@@ -0,0 +1,67 @@
+    
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        navigationController?.setNavigationBarHidden(true, animated: true)
```
계속 컨벤션이 왔다갔다 하고있죠?ㅎㅎ
그리고 NavigationBarHidden하는 주체가 ListView로 하는것도 가능할 것 같은데요. 
혹은 hidden을 WillDisappear에서 할 수도 있지 않을까요?
어떻게 생각하세요??🥸

dnwhd0112 replied:
1. hidden을 ListView 주체로 할 경우 처음 화면에서 네비게이션 만큼의 공백이 일어나서 이는 좋지 못하다고 느껴졌습니다.
2. 다른 viewController에서 제어하는게 좋아 보이지 않아서 일단 기존 코드를 유지시키는게 더 맞다고 생각했습니다.

lina0322 commented on Expo1900/MainViewController.swift:
```diff
@@ -0,0 +1,67 @@
+        guard let heritageListVC = storyboard?.instantiateViewController(withIdentifier: IdentifierName.heritageListViewController) as? HeritageListViewController else {
+            return
+        }
+        self.navigationController?.pushViewController(heritageListVC, animated: true)
```
왜 self가 필요한지, 혹은 왜 명시했는지에 대한 이유들이 명확하지 않으면
서로 얘기를 했더라도 헷갈리고 적용이 안 될 수 있어요.
코드를 짤 때, 이해한것같아도 이해 못한게 생각보다 많아요.ㅎㅎ 많은 대화가 필요합니당

dnwhd0112 replied:
의논 결과 다음과 같이 하기로 했습니다.
- self는 모든 인스턴스들이 암시적으로 생성하는 프로퍼티이다. 인스턴스 내부의 프로퍼티와 변수가 이름이 같은 경우, 구분을 위해 프로퍼티 앞에 self를 붙인다

lina0322 commented on Expo1900/MainViewController.swift:
```diff
@@ -0,0 +1,67 @@
+// MARK: - IBAction method
+extension MainViewController {
+    @IBAction func showListButtonIsTapped(_ sender: UIButton) {
+        guard let heritageListVC = storyboard?.instantiateViewController(withIdentifier: IdentifierName.heritageListViewController) as? HeritageListViewController else {
```
아니 이게 무슨 일이에요ㅠㅠ 
VC 무엇이죠..?😰

dnwhd0112 replied:
헉... 바로 바꾸도록 할게요!

lina0322 commented on Expo1900/MainViewController.swift:
```diff
@@ -0,0 +1,67 @@
+// MARK: - logic method
+extension MainViewController {
+    private func loadInformation() -> ExpositionInfo? {
+        let jsonManager: JsonManagerable = JsonManager()
```
JsonManagerable 타입은 왜 명시했나요??

dnwhd0112 replied:
JsonManager를 할당하는게 아닌 JsonManagerable 프로토콜을 만족하는 인스턴스만 받기 위해서 입니다. 이렇게 해놓으면 추후에 JsonManager를 다른 인스턴스로 변경하였을때 코드를 바꾸는 일이 적어질것이라고 생각해서 이렇게 구현하였습니다.

lina0322 replied:
오.. 잘 모르겠네요ㅎㅎ;; 일단 이대로 가보죠

lina0322 commented on Expo1900/MainViewController.swift:
```diff
@@ -0,0 +1,67 @@
+    }
+}
+
+// MARK: - logic method
```
logic method라는게 어떤 의미를 가지고있나용?

dnwhd0112 replied:
해당 타입 내부에서 쓰이는 메소드라는 의미를 가지고 있습니다! IBOutlet, IBAction, extension등 해당되지 않는 것들에 대한것만 모아놓는걸로 생각해서 만들었어요.

lina0322 commented on Expo1900/MainViewController.swift:
```diff
@@ -0,0 +1,67 @@
+    }
+    
+    private func splitTitleToTwoLines(from title: String) -> String {
+        return title.replacingOccurrences(of: "(", with: "\n(")
```
사실 이런 줄바꿈같은건, 기획에서 줄을 바꾸고싶다하면 정해줘서 서버에서 변경해서 내려줘야겠죠?ㅎㅎ
하지만 지금 상황으로서는 가장 좋은 시도였다고 생각해요!!

lina0322 commented on Expo1900/MainViewController.swift:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    private func setTextOfMainView(_ expositionInfo: ExpositionInfo) {
+        titleLabel.text = splitTitleToTwoLines(from: expositionInfo.title)
+        visitorLabel.text = "방문객 : \(expositionInfo.visitors) 명"
+        locationLabel.text = "개최지 : \(expositionInfo.location)"
+        durationLabel.text = "개최 기간 : \(expositionInfo.duration)"
+        descriptionLabel.text = expositionInfo.description
```
혹시 String의 format에 대해서 알고있나요? 어떤 차이가 있을까요?👀

dnwhd0112 replied:
format 이 좀 더 다양한 타입(integer, float) 을 String 형식으로 수정하기 편한거 같아요!

lina0322 reviewed (changes requested):
안녕하세요, 두 분!
이번 스텝도 열심히 진행하고 계시는군요!!!
항상 노력하는 모습 좋습니다~

---
### 2. 스토리보드 삽질
어떤게 문제였고 어떻게 해결했나요. 리드미에 정리해주세요.. 궁그매요...

### 3. data 전달
저는 첫번째 방식을 더 선호하는 편이긴 합니다ㅎㅎ
뭐든 장단점이 있는 것 같아요

### 5. UITableViewCell을 리턴
좋습니다. 저도 이렇게 해요~ 앱을 터트릴 순 없지 않겠습니까!!

--- 

추가로 네비게이션 왼쪽 선이 좀 짧은 것 같아요?!!ㅎㅎㅎ
![IMG_7074EA84285C-1](https://user-images.githubusercontent.com/49546979/163756870-6143b165-6177-41d2-a8cb-39eaaeab5106.jpeg)

코멘트 남겼으니까 천천히 보고, 또 열심히 해봅시다!!

marisol-develop committed:
refactor: self 컨벤션 수정

marisol-develop committed:
del: awakeFromNib, setSelected 메서드 삭제

marisol-develop committed:
refactor: self 컨벤션 수정

dnwhd0112 committed:
refactor: 배열 안전하게 접근하도록 변경, HeritageListViewController 초기화 함수 구현

extension Array의 subscript 추가 및 array 접근시 guard let 으로 접근
setInitialSetting 값 초기화 함수 구현

dnwhd0112 committed:
refactor: 테스트 객체 제거

dnwhd0112 committed:
refactor: 에러 처리 방법 변경

print -> alert 창을 띄워주는 방식으로 변경

marisol-develop committed:
rename: decodedResult로 메서드명 수정

dnwhd0112 committed:
refactor: 에러 메서지 protocal로 처리하도록 변경, heritageListVC 축약 제거

marisol-develop committed:
refactor: identifier 설정 방법 변경

lina0322 commented on Expo1900/HeritageListViewController.swift:
```diff
@@ -0,0 +1,90 @@
+
+class HeritageListViewController: UIViewController, GenerateErrorAlertProtocol {
+
+    @IBOutlet private var heritageListTableView: UITableView!
```
weak 키워드가 여긴 없고, 다른 곳에는 있네요. 어떤 부분으로 차이점을 두었나용?

lina0322 commented on Expo1900/HeritageListViewController.swift:
```diff
@@ -0,0 +1,90 @@
+
+    @IBOutlet private var heritageListTableView: UITableView!
+    private var items: [Item] = []
+    static let heritageListViewControllerName = String(describing: self)
```
한 단계 나아졌네요ㅎㅎ

lina0322 commented on Expo1900/HeritageListCell.swift:
```diff
@@ -0,0 +1,21 @@
+    @IBOutlet private weak var heritageTitleLabel: UILabel!
+    @IBOutlet private weak var heritageShortDescriptionLabel: UILabel!
+    
+    func setContentOfCell(_ item: Item) {
```
item을 명시하고 넘겨주는게 더 좋을 것 같아보이긴해요!

lina0322 commented on Expo1900/Extension/ArrayExtension.swift:
```diff
@@ -0,0 +1,12 @@
+//  Created by 우롱차, marisol on 2022/04/19.
+//
+
+extension Array {
+    public subscript(safe index: Int) -> Element? {
+        return indices ~= index ? self[index] : nil
+    }
+}
```
LGTM👍

lina0322 commented on Expo1900/ExpoError.swift:
```diff
@@ -2,10 +2,32 @@
     case noFileError
 }
+
+protocol GenerateErrorAlertProtocol {
+    func makeUIAlertControllerFromError(_ error: Error) -> UIAlertController
+}
```
Good👏

lina0322 commented on Expo1900/HeritageListViewController.swift:
```diff
@@ -0,0 +1,90 @@
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let identifier = String(describing: HeritageListCell.self)
```
HeritageListCell이 identifier를 들고있으면 더 좋겠네요ㅎㅎ

lina0322 commented on Expo1900/HeritageListViewController.swift:
```diff
@@ -0,0 +1,90 @@
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let identifier = String(describing: HeritageListCell.self)
+
+        guard let cell = heritageListTableView.dequeueReusableCell(withIdentifier: identifier, for: indexPath) as? HeritageListCell else {
+            return UITableViewCell()
+        }
+        
+        guard let item = items[safe: indexPath.row] else {
+            return UITableViewCell()
+        }
```
guard를 나눠서 작성한 이유가 있나요?

lina0322 commented on Expo1900/JsonManager.swift:
```diff
@@ -8,12 +8,12 @@
 
 struct JsonManager: JsonManagerable {
-    func decodedItems() throws -> [Item] {
+    func decodedResult() throws -> [Item] {
```
이름이 잘 이해가지 않는데 어떤의미로 사용했나용?ㅎㅎ 이전것도 약간 다소 애매한데, 흠..

lina0322 commented on Expo1900/MainViewController.swift:
```diff
@@ -0,0 +1,70 @@
+
+import UIKit
+
+class MainViewController: UIViewController, GenerateErrorAlertProtocol {
```
final 키워드 공부해보면 좋을거같아요~
당장은 아니어도 천천히🙏

lina0322 commented on Expo1900/MainViewController.swift:
```diff
@@ -0,0 +1,70 @@
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        guard let expositionInfo = loadInformation() else {
+            return
+        }
+        setTextOfMainView(expositionInfo)
```
guard로 return하면 하단 함수는 실행되지 않는데, 그걸 의도한 것이라면 `setTextOfMainView`함수 내부에서 작업하는게 좋지 않았을까요?

lina0322 commented on Expo1900/DetailViewController.swift:
```diff
@@ -0,0 +1,33 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        setContentsOfDetailView()
+        title = item?.name
```
요 title은 별도로 둘건가용?👀

lina0322 reviewed (approved):
코멘트 추가로 남겼어요! 
수정을 지금 하고 넘어가도 좋고, 다음 step에서 함께 하겠다고 하시면 머지 해드릴게요🙏

dnwhd0112 committed:
refactor: step2 리펙토링

class앞에 final 추가, identifier 가져오는 방식 변경 -> 각 viewController 마다 static프로퍼티 identifier추가. 일부 함수 구현부분 조정

marisol-develop committed:
feat: MainView 화면만 세로로 고정하도록 구현

dnwhd0112 committed:
feat: 오토 레이아웃 수정

dnwhd0112 committed:
refactor: 스토리 보드 수정 및 불필요한 코드 삭제

marisol-develop committed:
feat: dynamic type 적용

dnwhd0112 committed:
feat: storyBoard 변경 및 화면 회전 처리 추가

lina0322 commented:
머지하고 다음 스텝에서 함께 진행하기로 했습니다🙏
